implement `dao_migrator` module

Art3miX opened the pull request:
[Discussion](https://github.com/DA0-DA0/dao-contracts/discussions/607)

`dao_migrator` module which do the module migrations and test it went successfully.

Probably need more tests? I think I covered the basic tests that is handled by this contract.

Key changes from different discussions:
1. I removed the `Conjunction` ExecuteMsg, this is not needed as `ExecuteProposalHook` is doing the same for us already.
3. I was planning on adding `dao_core` migration tests, but because `dao_migrator` doesn't handle `dao_core` migration and relies on it to already be v2, we can't do those tests (because the migrator is initialized after core is at v2 already)

0xekez reviewed (commented):
honestly all i have to say: god tier.

i have very few comments, and none of them are on correctness. amazing work.

let me know when the tests are ready for review!

Art3miX commented on contracts/external/dao-migrator/src/msg.rs:
```diff
@@ -0,0 +1,30 @@
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub migration_params: MigrationParams,
+    pub sub_daos: Option<Vec<SubDao>>,
```
I do agree that empty vec gives us the same result, but in terms of UI or ICS721 (public interface) it makes sense to have optional fields, if some chain doesn't have any reference to `class_data`, why should they include it just to talk with our CW contract? they can just cleanly continue to ignore it.

For DAO DAO I do agree not optional is better.

Art3miX commented on contracts/external/dao-migrator/src/msg.rs:
```diff
@@ -0,0 +1,30 @@
+pub struct InstantiateMsg {
+    pub migration_params: MigrationParams,
+    pub sub_daos: Option<Vec<SubDao>>,
+    pub v1_code_ids: V1CodeIds,
+    pub v2_code_ids: V2CodeIds,
```
I was thinking about it, obviously it does make sense, but I think the byte cost is worth it:
1. It makes it clear what you expect/use in code, and makes understanding it much easier
2. in the future, you will be able to just add V3 code ids struct, add some fields to v2 code id and you are done in that area.

I don't have strong feelings about it, so just let me know the final call

0xekez replied:
i'm convinced by your point about v3 and clearly delineating the namespace of the code IDs. thanks!

Art3miX commented on contracts/external/dao-migrator/src/msg.rs:
```diff
@@ -0,0 +1,30 @@
+}
+
+#[cw_serde]
+pub enum ExecuteMsg {
+    /// Detects the current DAO configuration and performs a migration
+    /// checking state before and after to smoke test the migration's
+    /// success. This module will remove itself on this message's
+    /// completion regardless of the migration's success.
+    MigrateV1ToV2 {
+        sub_daos: Option<Vec<SubDao>>,
+        params: MigrationParams,
+        v1_code_ids: V1CodeIds,
+        v2_code_ids: V2CodeIds,
+    },
+}
```
I feel like if we do it, we turn this contract to `dao_migrator_v2` instead of a migrator.
so continue to my previous comment, should this be 0 future thinking?

0xekez replied:
fair point. this is your contract, so we can go with your rules here. :) personally, i tend to optimize robustness > future proof whenever those two are being compared. it's also hard to know if there even will be a DAO DAO v3! i would like v2 to be the last version tbh.

Art3miX commented on contracts/external/dao-migrator/src/types.rs:
```diff
@@ -0,0 +1,125 @@
+    pub proposal_counts: Vec<u64>,
+    pub proposals: Vec<dao_proposal_single::proposal::SingleChoiceProposal>,
+    pub total_voting_power: Uint128,
+    pub single_voting_power: Uint128,
```
This is the voting power of the proposer of the last proposal, added comment.

Art3miX commented on contracts/external/dao-migrator/src/utils/state_queries.rs:
```diff
@@ -0,0 +1,261 @@
+                )
+                .unwrap();
+
+            // TODO: What happens when there is no proposals on the module? we should ignore testing it
+            // and handle it as "tested?"
+            let proposal = proposals.proposals.first().unwrap().proposal.clone();
```
Feels wrong to bail out in this case, the DAO doesn't have proposal modules so he can't migrate other modules? 
Can you even have a functioning DAO with voting module but without proposal module? if that doesn't make sense then yea, I agree with you, but if its possible to have a functioning DAO without custom modules, and not have a proposal module, then this should not bail out.

Also V1 doesn't have subDaos, so its not possible to do this migration from a subDao, right? unless you first migrate core to V2 without migrator, add subDaos and then execute migrator directly.

0xekez replied:
> Also V1 doesn't have subDaos, so its not possible to do this migration from a subDao, right?

great question. v1 does have subDAOs, it just doesn't have the ability for parent DAO's to "recognize" subDAOs.

> Can you even have a functioning DAO with voting module but without proposal module?

nope, you always need a proposal module. in this case what i'm commenting on is the state where there is a proposal module, but there is no proposal in that module.

> if its possible to have a functioning DAO without custom modules, and not have a proposal module, then this should not bail out.

fair point. thanks for pushing back on these. i agree.

Art3miX replied:
I think bailing out make sense here, this data is needed for other tests, and if we don't have it, our tests are not complete.

Art3miX commented on contracts/external/dao-migrator/src/utils/state_queries.rs:
```diff
@@ -0,0 +1,261 @@
+            // and handle it as "tested?"
+            let proposal = proposals.proposals.first().unwrap().proposal.clone();
+
+            if last_proposal.is_none() {
```
I like `sample_proposal`, changed.

Art3miX commented on contracts/external/dao-migrator/src/utils/state_queries.rs:
```diff
@@ -0,0 +1,261 @@
+) -> StdResult<Uint128> {
+    let res: dao_interface::voting::VotingPowerAtHeightResponse = deps.querier.query_wasm_smart(
+        voting_addr,
+        &dao_voting_cw20_staked::msg::QueryMsg::VotingPowerAtHeight {
```
My idea was to use specifically what is used in production, there are some queries that wasn't changed, so I could use the same queries, but for good measure I want to use the one that is used by the contract we are calling.
So I was first checking if the contract have what I need, and if not only then I went to look in `dao_interface`, should it be the opposite?

0xekez replied:
great point as always. for this case i think you are right, this is an exception to the rule :)

Art3miX commented on contracts/external/dao-migrator/src/contract.rs:
```diff
@@ -0,0 +1,402 @@
+    });
+
+    if !success_proposal {
+        Err(proposal_error.unwrap())
```
So I was actually looking into that example, and if I have some simple map like in there, I have no problems with it, but for some reason I didn't manage to do it with a more complex map and custom errors (ContractError and not StdError).

Also I think I can't do a map here, because I need to ignore one of the entries of the vec, `if env.contract.address == module.address` we need to ignore the migrator module here.
Plus, I don't actually care about the vec after, just want to loop over it.

0xekez replied:
> I need to ignore one of the entries of the vec, if env.contract.address == module.address

`iter().filter(|m| m.address != env.contract.address).map(/* ... */)`

my particular concern here is the interplay between `success_proposal` and `proposal_error` that is not expressed in the type system (and thus the need for the unwrap). if we, in the future, modify this contract and forget to correctly handle that interplay, that's a great way to have a bug.

i _think_ that using the collect approach we can move the management of that interplay to the rust standard library which is a much nicer place for it.

> for some reason I didn't manage to do it with a more complex map and custom errors (ContractError and not StdError)

this is definitely tricky. could you DM me the error that you get? i can help you fix this. I run into this often.

Art3miX replied:
Ha, some how I managed to do it on the first try, no idea why I was struggling with it at the first place, pretty straight forward honestly.

I do want to raise 2 questions about it:

1. I didn't do the filter, because I don't care about the vec afterward, so instead I just ignore our module, I just Ok(()) if it is the migrator module, that way, we only do 1 loop instead of 2.
2. from the `.map()` docs: `If you're doing some sort of looping for a side effect, it's considered more idiomatic to use for than map()`, how important is it in this case? seems like this sentence is talking exactly about my case here.

Art3miX replied:
Should have ran clippy before commenting (future note), clippy complained about it, and suggested to use `.try_for_each()` which make sense.

Art3miX commented:
Awesome review @0xekez thank you very much!

I resolved what I changed, and left unresolved stuff I still have questions about.

0xekez commented:
@Art3miX i left comments on all your responses. thank you for pushing back on those. i like your arguments.

0xekez commented on contracts/external/dao-migrator/Cargo.toml:
```diff
@@ -0,0 +1,52 @@
+
+[dependencies]
+cosmwasm-std     = { workspace = true, features = ["ibc3"] }
+cosmwasm-storage = { workspace = true }
```
```suggestion
```

0xekez commented on contracts/external/dao-migrator/src/contract.rs:
```diff
@@ -0,0 +1,388 @@
+    migration_params: MigrationParams,
+    v1_code_ids: V1CodeIds,
+    v2_code_ids: V2CodeIds,
+) -> Result<Response, ContractError> {
```
could we check `info.sender == CORE_ADDR.load(deps.storage)` here? so long as this is atomic, this shouldn't matter, but having the check makes this safer in the event that we mess up atomicity.

0xekez commented on contracts/external/dao-migrator/src/testing/test_migration.rs:
```diff
@@ -0,0 +1,147 @@
+
+#[test]
+fn test_execute_migration() {
+    let (mut app, module_addrs, v1_code_ids) = setup_dao_v1(VotingType::Cw20);
```
is there a version of this test with `VotingType::Cw4`? that would be nice to have. great work on that instantiation abstraction btw.

0xekez commented on contracts/external/dao-migrator/src/testing/test_migration.rs:
```diff
@@ -0,0 +1,147 @@
+    //NOTE: We add 1 to count because we create a new proposal in execute_migration
+    test_state_v1.proposal_count += 1;
+
+    execute_migration(app.borrow_mut(), &module_addrs, v1_code_ids, None).unwrap();
```
could we also query and assert after this migration that the module is no longer one of the proposal modules?

0xekez reviewed (approved):
amazing work. i left a few small comments, but i think this is good to go.

as a last thing @Art3miX how would you feel about modifying the `dao-core` migrate method to allow adding this contract as part of migration? i think this can be done by taking `MigrateV1ToV2 ` and a code ID as part of migration (optionally), and if it is present dispatching a message to add this module once migration completes.

0xekez commented on contracts/external/dao-migrator/src/testing/test_migration.rs:
```diff
@@ -1,36 +1,70 @@
+
+#[test]
+fn test_execute_migration() {
+    basic_test(VotingType::Cw20);
+
+    basic_test(VotingType::Cw4);
```
amazing 🙏

0xekez reviewed (approved):
lg2m. thank you! i like how this turned out.

JakeHartnell reviewed (approved):
LFG!

JakeHartnell commented:
Just failing lints.

Art3miX commented:
I added `rm -rf ./schema/raw` into the schema script, as suggested [here](https://github.com/CosmWasm/ts-codegen/issues/90#issuecomment-1409674460).

Also updated rustup version to latest which fixed my local clippy error, hopefully CI should pass now.

Art3miX commented:
We have an issue with fmt in `cw-denom`: https://github.com/DA0-DA0/dao-contracts/blob/main/packages/cw-denom/src/lib.rs#L302

I think `cargo fmt` ignores unicode for its comment spacing, the CI do a fmt check, which returns error because of the spacing.
